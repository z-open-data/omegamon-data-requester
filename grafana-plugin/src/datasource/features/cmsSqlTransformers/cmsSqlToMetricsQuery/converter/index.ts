export { convertToMetricsQuery } from './convertToMetricsQuery';
export type { ConversionProblem } from './internals';
export { TokenMapping, ConverterError } from './internals';
