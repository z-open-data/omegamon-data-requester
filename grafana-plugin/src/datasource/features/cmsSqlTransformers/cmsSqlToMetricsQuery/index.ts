export { detectApplicationAndTable } from './detectApplicationAndTable';
export { parseCmsSqlQuery } from './parseCmsSqlQuery';
export type { CmsSqlParsingProblem, CmsSqlParsingResult } from './parseCmsSqlQuery';
export { buildLinesMap, getTokensPosition } from './SubstringPosition';
export { queries } from './snapshotTestCases';
