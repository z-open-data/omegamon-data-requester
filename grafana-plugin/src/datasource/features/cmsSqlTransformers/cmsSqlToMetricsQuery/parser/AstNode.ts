import {
  TokenOf,
  TokenOfGenericKind,
  Value,
} from 'datasource/features/cmsSqlTransformers/cmsSqlToMetricsQuery/tokenizer';

export interface UnaryExpression {
  nodeType: 'unary';
  operator: TokenOf<'add_arithmetic_operator' | 'sub_arithmetic_operator'>;
  childExpr: Expression;
}

export interface BinaryExpression {
  nodeType: 'binary';
  left: Expression;
  right: Expression;
  operator: TokenOfGenericKind<'arithmetic_operator' | 'comparison_operator' | 'condition_connector'>;
}

export interface UnqualifiedIdentifier {
  nodeType: 'unqualified_identifier';
  token: TokenOf<'identifier'>;
}

export interface QualifiedIdentifier {
  nodeType: 'qualified_identifier';
  left: TokenOf<'identifier'>;
  right: TokenOf<'identifier'>;
}

export interface ValueExpression<T extends Value = Value> {
  nodeType: 'value';
  tokens: T[];
  asUserDefinedValue: string | number;
}

export type Identifier = UnqualifiedIdentifier | QualifiedIdentifier;

export interface FunctionCall {
  nodeType: 'function_call';
  funcIdentifier: Identifier;
  openParenthesis: TokenOf<'open_parenthesis'>;
  closeParenthesis: TokenOf<'close_parenthesis'>;
  args: Expression[];
}

export type Expression = ValueExpression | UnaryExpression | BinaryExpression | FunctionCall | Identifier;

export type SelectItem = Identifier | FunctionCall;

export type SelectClause = {
  nodeType: 'select';
  items: SelectItem[];
  keyword: TokenOf<'select_keyword'>;
};

export type FromClause = {
  nodeType: 'from';
  tables: Identifier[];
  keyword: TokenOf<'from_keyword'>;
};

export type AtClause = {
  nodeType: 'at';
  items: Expression[];
  keyword: TokenOf<'at_keyword'>;
};

export type HistoryClause = {
  nodeType: 'history';
  keyword: TokenOf<'history_keyword'>;
};

export type WhereClause = {
  nodeType: 'where';
  filter: Expression;
  keyword: TokenOf<'where_keyword'>;
};

export type HavingClause = {
  nodeType: 'having';
  filter: Expression;
  keyword: TokenOf<'having_keyword'>;
};

export type GroupByClause = {
  nodeType: 'group_by';
  columns: Identifier[];
  keywords: [TokenOf<'group_keyword'>, TokenOf<'by_keyword'>];
};

export type OrderByExpression = {
  nodeType: 'order_by_expression';
  identifier: Identifier;
  direction: TokenOfGenericKind<'order_direction'> | null;
};

export type OrderByClause = {
  nodeType: 'order_by';
  items: OrderByExpression[];
  keywords: [TokenOf<'order_keyword'>, TokenOf<'by_keyword'>];
};

export type SelectStatement = {
  nodeType: 'select_statement';
  select: SelectClause;
  from: FromClause;
  at: AtClause | null;
  history: HistoryClause | null;
  where: WhereClause | null;
  groupBy: GroupByClause | null;
  having: HavingClause | null;
  orderBy: OrderByClause | null;
};

export type AstNode =
  | Expression
  | SelectClause
  | FromClause
  | AtClause
  | HistoryClause
  | WhereClause
  | GroupByClause
  | HavingClause
  | OrderByExpression
  | OrderByClause
  | SelectStatement;
