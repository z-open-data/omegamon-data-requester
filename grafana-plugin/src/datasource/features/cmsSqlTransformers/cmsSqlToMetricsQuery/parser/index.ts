export * from './AstNode';
export * from './helpers';
export { parseSelectStatement } from './parseSelectStatement';
export { skipUntil } from './common';
export { parseFromClause } from './parseFromClause';
export { ParserError } from './ParserError';
