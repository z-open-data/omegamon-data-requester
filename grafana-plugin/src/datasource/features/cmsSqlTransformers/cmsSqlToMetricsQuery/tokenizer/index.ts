export { getVariablesFromStr } from './processStringWithVariables';
export { isValueGroupToken, isValueToken } from './token';
export type {
  ResolvedDecimalLiteral,
  ResolvedHexadecimalLiteral,
  StringLiteral,
  Token,
  TokenOf,
  TokenBase,
  TokenOfGenericKind,
  UnresolvedDecimalLiteral,
  UnresolvedHexadecimalLiteral,
  Value,
  Variable,
} from './token';
export { KEYWORD_LIKE_KINDS, KNOWN_CMS_SQL_KEYWORDS } from './tokenKind';
export type { GenericKind, TokenKind } from './tokenKind';
export { tokenize } from './tokenize';
