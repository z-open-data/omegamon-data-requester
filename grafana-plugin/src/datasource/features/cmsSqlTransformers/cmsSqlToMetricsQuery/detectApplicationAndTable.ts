import { skipUntil, parseFromClause } from './parser';
import { tokenize } from './tokenizer';

export function detectApplicationAndTable(cmsSqlSource: string): { applicationCode: string; tableId?: string } | null {
  try {
    const tokens = tokenize(cmsSqlSource);
    const [_, start] = skipUntil(tokens, 'from_keyword');
    const [from] = parseFromClause(start, true);
    if (!from) {
      return null;
    }

    const tableIdentifier = from.tables[0];
    if (!tableIdentifier) {
      return null;
    }

    if (tableIdentifier.nodeType !== 'qualified_identifier') {
      return {
        applicationCode: tableIdentifier.token.tokenText.toUpperCase(),
      };
    }
    return {
      applicationCode: tableIdentifier.left.tokenText.toUpperCase(),
      tableId: tableIdentifier.right.tokenText.toUpperCase(),
    };
  } catch (_) {
    return null;
  }
}
