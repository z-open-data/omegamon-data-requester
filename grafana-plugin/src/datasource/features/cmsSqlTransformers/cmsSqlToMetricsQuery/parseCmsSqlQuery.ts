import { throwIfNullish } from 'public-common';

import { MetricsQueryParams } from 'datasource/domain';
import { MetadataLoader } from 'datasource/features/metadata';
import { validateMetricsQuery } from 'datasource/features/validation/validateMetricsQuery';

import { ConversionProblem, convertToMetricsQuery } from './converter';
import { getNodeTokens, parseSelectStatement } from './parser';
import { SubstringPosition, LinesMap, getTokensPosition, buildLinesMap, getTokenPosition } from './SubstringPosition';
import { tokenize } from './tokenizer';

export type CmsSqlParsingProblem = {
  severity: 'warning' | 'error' | 'informational';
  stage: 'tokenizer' | 'parser' | 'converter' | 'validator';
  message: string;
  position: SubstringPosition;
};

export type CmsSqlParsingResult = {
  queryObject: MetricsQueryParams | null;
  queryString: string;
  problems: CmsSqlParsingProblem[];
};

function processConvertionProblems(
  conversionProblems: ConversionProblem[],
  linesMap: LinesMap
): CmsSqlParsingProblem[] {
  const problems: CmsSqlParsingProblem[] = [];
  conversionProblems.forEach((problem) => {
    const { message, severity } = problem;
    if ('affectedAstNodes' in problem) {
      problem.affectedAstNodes.forEach((affectedAstNode) => {
        problems.push({
          severity,
          stage: 'converter' as const,
          message,
          position: getTokensPosition(getNodeTokens(affectedAstNode), linesMap),
        });
      });
    }

    if ('affectedTokens' in problem) {
      problems.push({
        severity,
        stage: 'converter' as const,
        message,
        position: getTokensPosition(problem.affectedTokens, linesMap),
      });
    }
  });

  return problems;
}

export async function parseCmsSqlQuery(
  queryString: string,
  metadataLoader: MetadataLoader
): Promise<CmsSqlParsingResult> {
  const linesMap = buildLinesMap(queryString);
  const tokens = tokenize(queryString);
  const allProblems = tokens.reduce((acc, token) => {
    const problems = token.problems.map((problem) => ({
      severity: problem.severity,
      stage: 'tokenizer' as const,
      message: problem.message,
      position: getTokenPosition(token, linesMap),
    }));
    acc.push(...problems);
    return acc;
  }, [] as CmsSqlParsingProblem[]);

  const { statement, problems: parsingProblems } = parseSelectStatement(tokens);
  allProblems.push(
    ...parsingProblems.map(({ severity, message, problemTokens }) => ({
      severity,
      stage: 'parser' as const,
      message,
      position: getTokensPosition(problemTokens, linesMap),
    }))
  );

  if (!statement) {
    return {
      queryString,
      queryObject: null,
      problems: allProblems,
    };
  }

  const {
    query: queryObject,
    problems: conversionProblems,
    getTokensOf,
  } = await convertToMetricsQuery(statement, metadataLoader);
  allProblems.push(...processConvertionProblems(conversionProblems, linesMap));

  if (queryObject) {
    (await validateMetricsQuery(queryObject, metadataLoader, getTokensOf)).forEach(
      ({ severity, message, affectedTokens }) => {
        throwIfNullish(
          affectedTokens,
          'validateMetricsQuery should always return affected tokens if getTokensOf was provided'
        );
        allProblems.push({
          severity: severity,
          stage: 'validator' as const,
          message,
          position: getTokensPosition(affectedTokens, linesMap),
        });
      }
    );
  }

  return {
    queryString,
    queryObject,
    problems: allProblems,
  };
}
