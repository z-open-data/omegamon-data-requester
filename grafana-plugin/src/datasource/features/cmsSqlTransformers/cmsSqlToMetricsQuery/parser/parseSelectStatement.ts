import { Token, TokenKind } from 'datasource/features/cmsSqlTransformers/cmsSqlToMetricsQuery/tokenizer';

import { AstNode, SelectStatement } from './AstNode';
import { BlockParser, skipUntil } from './common';
import { parseAtFuncCall } from './parseAtFuncCall';
import { parseWhereClause, parseHavingClause } from './parseFilterGroup';
import { parseFromClause } from './parseFromClause';
import { parseGroupByClause } from './parseGroupByClause';
import { parseHistoryFuncCall } from './parseHistoryFuncCall';
import { parseOrderByClause } from './parseOrderByClause';
import { ParserError } from './ParserError';
import { parseSelectClause } from './parseSelectClause';

export type ParsingProblem = {
  severity: 'warning' | 'error';
  message: string;
  problemTokens: Token[];
};

function parseClause<T extends AstNode>(
  tokens: Token[],
  blockParser: BlockParser<T>,
  syncToken: TokenKind
): [clause: T | null, problems: ParsingProblem[], rest: Token[]] {
  const problems: ParsingProblem[] = [];
  try {
    let start = tokens;

    const [skippedTokens, newStart] = skipUntil(tokens, syncToken);
    start = newStart;

    if (skippedTokens.length) {
      problems.push({
        severity: 'error',
        message: 'Unexpected token(s) found',
        problemTokens: skippedTokens,
      });
    }

    const [clause, rest] = blockParser(start);

    return [clause, problems, rest];
  } catch (e) {
    if (e instanceof ParserError) {
      problems.push({
        severity: 'error',
        message: e.message,
        problemTokens: e.tokens,
      });
    }
    return [null, problems, tokens];
  }
}

// It's basically a definition of select statement.
// Items in this tuple represent the order in which clauses are expected to
// be in the query (and in which parser will try to parse them).
// Each tuple item contains 2 elements: parser function and sync token kind.
// Parser function is simple: it's a function that parses corresponding clause.
// Sync token kind is interesting one. One of the goals of parser is to provide
// error handling. We want parser to provide errors as helpful and as complete
// as possible. So, once we failed to parse some clause, we still have to try
// to parse the rest. This is where syncing tokens are helpful. They define
// the keyword we want particular clause to be synced with. Once we failed
// to parse, let's say, from clause, we still will find the beginning of
// where clause and parse it successfully (or report new problems)
const SELECT_STATEMENT_DEFINITION = [
  [parseSelectClause, 'select_keyword'],
  [parseFromClause, 'from_keyword'],
  [parseAtFuncCall, 'at_keyword'],
  [parseHistoryFuncCall, 'history_keyword'],
  [parseWhereClause, 'where_keyword'],
  [parseGroupByClause, 'group_keyword'],
  [parseHavingClause, 'having_keyword'],
  [parseOrderByClause, 'order_keyword'],
] as const satisfies ReadonlyArray<readonly [BlockParser, TokenKind]>;

type ExtractBlockType<T extends BlockParser> = T extends BlockParser<infer R> ? R | null : never;
type GetBlocksTuple<T extends ReadonlyArray<readonly [BlockParser, TokenKind | null]>> = {
  [K in keyof T]: ExtractBlockType<T[K][0]>;
};
type BlocksTuple = GetBlocksTuple<typeof SELECT_STATEMENT_DEFINITION>;

function runParsers(tokens: Token[]): [blocks: BlocksTuple, problems: ParsingProblem[], rest: Token[]] {
  const result: Array<AstNode | null> = [];
  const problems: ParsingProblem[] = [];
  let currPos = tokens;

  for (const [blockParser, syncWith] of SELECT_STATEMENT_DEFINITION) {
    const [clause, blockProblems, rest] = parseClause(currPos, blockParser as BlockParser, syncWith);
    result.push(clause);
    problems.push(...blockProblems);
    currPos = rest;
  }

  return [result as unknown as BlocksTuple, problems, currPos];
}

export function parseSelectStatement(tokens: Token[]): {
  statement: SelectStatement | null;
  problems: ParsingProblem[];
} {
  const [[select, from, at, history, where, groupBy, having, orderBy], underlyingProblems, rest] = runParsers(tokens);

  if (!select || !from) {
    return {
      statement: null,
      problems: [
        ...underlyingProblems,
        {
          severity: 'error',
          message: 'Both SELECT and FROM clauses are mandatory',
          problemTokens: tokens,
        },
      ],
    };
  }

  const problems = [...underlyingProblems];
  const [endOfStatement, ...afterEnd] = rest;
  const afterTheStatement = endOfStatement?.kind === 'statement_separator' ? afterEnd : rest;
  if (afterTheStatement.length !== 0 && problems.length === 0) {
    // We don't want to report this problem if there are other problems already
    // detected because this one is way too generic
    problems.push({
      severity: 'error',
      message: 'Unexpected tokens at the end of statement',
      problemTokens: afterTheStatement,
    });
  }

  return {
    statement: {
      nodeType: 'select_statement',
      select,
      from,
      at,
      history,
      where,
      groupBy,
      having,
      orderBy,
    },
    problems,
  };
}
