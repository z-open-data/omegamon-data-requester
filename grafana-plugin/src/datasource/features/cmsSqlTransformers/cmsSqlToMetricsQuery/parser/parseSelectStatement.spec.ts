import { tokenize } from 'datasource/features/cmsSqlTransformers/cmsSqlToMetricsQuery/tokenizer';

import { SelectStatement } from './AstNode';
import { parseSelectStatement } from './parseSelectStatement';

test('all clauses in one query', () => {
  const { statement } = parseSelectStatement(
    tokenize(
      '\
          SELECT ORIGINNODE, O4ONLINE \
          FROM O4SRV.INODESTS \
          AT("*HUB") \
          HISTORY() \
          WHERE ORIGINNODE <> "Anything" \
          GROUP BY THRUNODE \
          HAVING AVG(O4ONLINE) = 5 \
          ORDER BY ORIGINNODE \
          '
    )
  );

  expect(statement).toMatchObject({
    select: {
      nodeType: 'select',
      items: [
        {
          nodeType: 'unqualified_identifier',
          token: {
            kind: 'identifier',
            tokenText: 'ORIGINNODE',
          },
        },
        {
          nodeType: 'unqualified_identifier',
          token: {
            kind: 'identifier',
            tokenText: 'O4ONLINE',
          },
        },
      ],
    },
    from: {
      nodeType: 'from',
      tables: [
        {
          nodeType: 'qualified_identifier',
          left: {
            kind: 'identifier',
            tokenText: 'O4SRV',
          },
          right: {
            kind: 'identifier',
            tokenText: 'INODESTS',
          },
        },
      ],
    },
    at: {
      nodeType: 'at',
      items: [
        {
          nodeType: 'value',
          tokens: [{ kind: 'string_literal', value: '*HUB' }],
        },
      ],
    },
  } as SelectStatement);
});

test('returns null if there is no select keyword match', () => {
  const { statement, problems } = parseSelectStatement(tokenize('ELECT ORIGINNODE, THRUNODE FROM O4SRV'));

  expect(statement).toBe(null);
  expect(problems).toMatchSnapshot();
});

test('returns partial match if comma is missed', () => {
  const _cmsSql = 'SELECT ORIGINNODE THRUNODE FROM O4SRV';
  const { statement, problems } = parseSelectStatement(tokenize(_cmsSql));

  expect({ _cmsSql, statement, problems }).toMatchSnapshot();
});

test('returns error if there is no identifier after the comma', () => {
  const _cmsSql = 'SELECT ORIGINNODE, 43 FROM O4SRV';
  const { statement, problems } = parseSelectStatement(tokenize(_cmsSql));

  expect(statement).toBeNull();
  expect(problems).toMatchObject([
    {
      message: 'Unexpected tokens after the separator',
    },
    {
      message: 'Unexpected token(s) found',
    },
    {
      message: 'Both SELECT and FROM clauses are mandatory',
    },
  ]);
});

test("Parse SELECT A FROM B.C WHERE A = 'str1${var1}''str2${var2}'", () => {
  const cmsSql = "SELECT A FROM B.C WHERE A = 'str1${var1}''str2${var2}'";
  const { problems, statement } = parseSelectStatement(tokenize(cmsSql));

  expect(problems).toHaveLength(0);
  expect(statement).toStrictEqual({
    at: null,
    from: {
      keyword: {
        endIdx: 13,
        genericKind: 'keyword',
        kind: 'from_keyword',
        problems: [],
        startIdx: 9,
        tokenText: 'FROM',
      },
      nodeType: 'from',
      tables: [
        {
          left: {
            endIdx: 15,
            genericKind: 'identifier',
            kind: 'identifier',
            problems: [],
            startIdx: 14,
            tokenText: 'B',
          },
          nodeType: 'qualified_identifier',
          right: {
            endIdx: 17,
            genericKind: 'identifier',
            kind: 'identifier',
            problems: [],
            startIdx: 16,
            tokenText: 'C',
          },
        },
      ],
    },
    groupBy: null,
    having: null,
    history: null,
    nodeType: 'select_statement',
    orderBy: null,
    select: {
      items: [
        {
          nodeType: 'unqualified_identifier',
          token: {
            endIdx: 8,
            genericKind: 'identifier',
            kind: 'identifier',
            problems: [],
            startIdx: 7,
            tokenText: 'A',
          },
        },
      ],
      keyword: {
        endIdx: 6,
        genericKind: 'keyword',
        kind: 'select_keyword',
        problems: [],
        startIdx: 0,
        tokenText: 'SELECT',
      },
      nodeType: 'select',
    },
    where: {
      filter: {
        left: {
          nodeType: 'unqualified_identifier',
          token: {
            endIdx: 25,
            genericKind: 'identifier',
            kind: 'identifier',
            problems: [],
            startIdx: 24,
            tokenText: 'A',
          },
        },
        nodeType: 'binary',
        operator: {
          endIdx: 27,
          genericKind: 'comparison_operator',
          kind: 'eq_comparison_operator',
          problems: [],
          startIdx: 26,
          tokenText: '=',
        },
        right: {
          asUserDefinedValue: "str1${var1}'str2${var2}",
          nodeType: 'value',
          tokens: [
            {
              endIdx: 33,
              genericKind: 'literal',
              kind: 'string_literal',
              problems: [],
              valueGroupId: {
                endIdx: 54,
                genericKind: 'literal',
                kind: 'string_literal',
                problems: [],
                startIdx: 28,
                tokenText: "'str1${var1}''str2${var2}'",
                value: "str1${var1}'str2${var2}",
              },
              startIdx: 28,
              tokenText: "'str1",
              value: 'str1',
            },
            {
              endIdx: 40,
              genericKind: 'variable',
              kind: 'variable',
              problems: [],
              valueGroupId: {
                endIdx: 54,
                genericKind: 'literal',
                kind: 'string_literal',
                problems: [],
                startIdx: 28,
                tokenText: "'str1${var1}''str2${var2}'",
                value: "str1${var1}'str2${var2}",
              },
              startIdx: 33,
              tokenText: '${var1}',
              varName: 'var1',
            },
            {
              endIdx: 46,
              genericKind: 'literal',
              kind: 'string_literal',
              problems: [],
              valueGroupId: {
                endIdx: 54,
                genericKind: 'literal',
                kind: 'string_literal',
                problems: [],
                startIdx: 28,
                tokenText: "'str1${var1}''str2${var2}'",
                value: "str1${var1}'str2${var2}",
              },
              startIdx: 40,
              tokenText: "''str2",
              value: "'str2",
            },
            {
              endIdx: 53,
              genericKind: 'variable',
              kind: 'variable',
              problems: [],
              valueGroupId: {
                endIdx: 54,
                genericKind: 'literal',
                kind: 'string_literal',
                problems: [],
                startIdx: 28,
                tokenText: "'str1${var1}''str2${var2}'",
                value: "str1${var1}'str2${var2}",
              },
              startIdx: 46,
              tokenText: '${var2}',
              varName: 'var2',
            },
            {
              endIdx: 54,
              genericKind: 'literal',
              kind: 'string_literal',
              problems: [],
              valueGroupId: {
                endIdx: 54,
                genericKind: 'literal',
                kind: 'string_literal',
                problems: [],
                startIdx: 28,
                tokenText: "'str1${var1}''str2${var2}'",
                value: "str1${var1}'str2${var2}",
              },
              startIdx: 53,
              tokenText: "'",
              value: '',
            },
          ],
        },
      },
      keyword: {
        endIdx: 23,
        genericKind: 'keyword',
        kind: 'where_keyword',
        problems: [],
        startIdx: 18,
        tokenText: 'WHERE',
      },
      nodeType: 'where',
    },
  });
});

test("Parse SELECT A FROM B.C WHERE A = 'str1${var1}' 'str2${var2}'", () => {
  const cmsSql = "SELECT A FROM B.C WHERE A = 'str1${var1}' 'str2${var2}'";
  const { problems } = parseSelectStatement(tokenize(cmsSql));

  expect(problems).toStrictEqual([
    {
      message: 'Unexpected tokens at the end of statement',
      problemTokens: [
        {
          endIdx: 47,
          genericKind: 'literal',
          kind: 'string_literal',
          problems: [],
          valueGroupId: {
            endIdx: 55,
            genericKind: 'literal',
            kind: 'string_literal',
            problems: [],
            startIdx: 42,
            tokenText: "'str2${var2}'",
            value: 'str2${var2}',
          },
          startIdx: 42,
          tokenText: "'str2",
          value: 'str2',
        },
        {
          endIdx: 54,
          genericKind: 'variable',
          kind: 'variable',
          problems: [],
          valueGroupId: {
            endIdx: 55,
            genericKind: 'literal',
            kind: 'string_literal',
            problems: [],
            startIdx: 42,
            tokenText: "'str2${var2}'",
            value: 'str2${var2}',
          },
          startIdx: 47,
          tokenText: '${var2}',
          varName: 'var2',
        },
        {
          endIdx: 55,
          genericKind: 'literal',
          kind: 'string_literal',
          problems: [],
          valueGroupId: {
            endIdx: 55,
            genericKind: 'literal',
            kind: 'string_literal',
            problems: [],
            startIdx: 42,
            tokenText: "'str2${var2}'",
            value: 'str2${var2}',
          },
          startIdx: 54,
          tokenText: "'",
          value: '',
        },
      ],
      severity: 'error',
    },
  ]);
});
