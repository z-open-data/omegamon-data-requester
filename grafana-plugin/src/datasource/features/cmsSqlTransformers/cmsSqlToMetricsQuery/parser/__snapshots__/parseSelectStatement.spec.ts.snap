// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns null if there is no select keyword match 1`] = `
[
  {
    "message": "Unexpected token(s) found",
    "problemTokens": [
      {
        "endIdx": 5,
        "genericKind": "identifier",
        "kind": "identifier",
        "problems": [],
        "startIdx": 0,
        "tokenText": "ELECT",
      },
      {
        "endIdx": 16,
        "genericKind": "identifier",
        "kind": "identifier",
        "problems": [],
        "startIdx": 6,
        "tokenText": "ORIGINNODE",
      },
      {
        "endIdx": 17,
        "genericKind": "separator",
        "kind": "comma_separator",
        "problems": [],
        "startIdx": 16,
        "tokenText": ",",
      },
      {
        "endIdx": 26,
        "genericKind": "identifier",
        "kind": "identifier",
        "problems": [],
        "startIdx": 18,
        "tokenText": "THRUNODE",
      },
    ],
    "severity": "error",
  },
  {
    "message": "Both SELECT and FROM clauses are mandatory",
    "problemTokens": [
      {
        "endIdx": 5,
        "genericKind": "identifier",
        "kind": "identifier",
        "problems": [],
        "startIdx": 0,
        "tokenText": "ELECT",
      },
      {
        "endIdx": 16,
        "genericKind": "identifier",
        "kind": "identifier",
        "problems": [],
        "startIdx": 6,
        "tokenText": "ORIGINNODE",
      },
      {
        "endIdx": 17,
        "genericKind": "separator",
        "kind": "comma_separator",
        "problems": [],
        "startIdx": 16,
        "tokenText": ",",
      },
      {
        "endIdx": 26,
        "genericKind": "identifier",
        "kind": "identifier",
        "problems": [],
        "startIdx": 18,
        "tokenText": "THRUNODE",
      },
      {
        "endIdx": 31,
        "genericKind": "keyword",
        "kind": "from_keyword",
        "problems": [],
        "startIdx": 27,
        "tokenText": "FROM",
      },
      {
        "endIdx": 37,
        "genericKind": "identifier",
        "kind": "identifier",
        "problems": [],
        "startIdx": 32,
        "tokenText": "O4SRV",
      },
    ],
    "severity": "error",
  },
]
`;

exports[`returns partial match if comma is missed 1`] = `
{
  "_cmsSql": "SELECT ORIGINNODE THRUNODE FROM O4SRV",
  "problems": [
    {
      "message": "Unexpected token(s) found",
      "problemTokens": [
        {
          "endIdx": 26,
          "genericKind": "identifier",
          "kind": "identifier",
          "problems": [],
          "startIdx": 18,
          "tokenText": "THRUNODE",
        },
      ],
      "severity": "error",
    },
  ],
  "statement": {
    "at": null,
    "from": {
      "keyword": {
        "endIdx": 31,
        "genericKind": "keyword",
        "kind": "from_keyword",
        "problems": [],
        "startIdx": 27,
        "tokenText": "FROM",
      },
      "nodeType": "from",
      "tables": [
        {
          "nodeType": "unqualified_identifier",
          "token": {
            "endIdx": 37,
            "genericKind": "identifier",
            "kind": "identifier",
            "problems": [],
            "startIdx": 32,
            "tokenText": "O4SRV",
          },
        },
      ],
    },
    "groupBy": null,
    "having": null,
    "history": null,
    "nodeType": "select_statement",
    "orderBy": null,
    "select": {
      "items": [
        {
          "nodeType": "unqualified_identifier",
          "token": {
            "endIdx": 17,
            "genericKind": "identifier",
            "kind": "identifier",
            "problems": [],
            "startIdx": 7,
            "tokenText": "ORIGINNODE",
          },
        },
      ],
      "keyword": {
        "endIdx": 6,
        "genericKind": "keyword",
        "kind": "select_keyword",
        "problems": [],
        "startIdx": 0,
        "tokenText": "SELECT",
      },
      "nodeType": "select",
    },
    "where": null,
  },
}
`;
